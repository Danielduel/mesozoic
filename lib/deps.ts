export {
  copy,
  emptyDir,
  ensureDir,
  walk,
} from "https://deno.land/std@0.153.0/fs/mod.ts";
export * as log from "https://deno.land/std@0.153.0/log/mod.ts";

export { sprintf } from "https://deno.land/std@0.153.0/fmt/printf.ts";

export {
  dirname,
  extname,
  globToRegExp,
  join,
  resolve,
  toFileUrl,
} from "https://deno.land/std@0.153.0/path/mod.ts";
