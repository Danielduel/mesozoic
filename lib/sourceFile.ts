import { dirname, ensureDir, extname, join, sprintf } from "./deps.ts";

export class SourceFile {
  private aliasPath?: string;
  private locked = true;

  /**
   * @param filePath The absolute filepath to the file.
   * @param rootPath The absolute filepath to the root of this file.
   */
  constructor(
    private readonly filePath: string,
    private readonly rootPath: string,
  ) {}

  path() {
    return this.filePath;
  }

  /**
   * @returns The relative path of this file to its root.
   */
  relativePath() {
    return this.path().replace(this.rootPath, ".");
  }

  alias() {
    return this.aliasPath;
  }

  relativeAlias() {
    return this.alias()?.replace(this.rootPath, ".");
  }

  root() {
    return this.rootPath;
  }

  getExtension() {
    return extname(this.path());
  }

  /**
   * @returns The content of the file as a string.
   */
  async read() {
    return new TextDecoder().decode(await this.readBytes());
  }

  write(content: string) {
    if (this.isLocked()) {
      throw new Error(
        sprintf("cannot write file because it is locked: %s", this.path()),
      );
    }
    return Deno.writeFile(this.path(), new TextEncoder().encode(content));
  }

  /**
   * @returns The content of the file as an array of bytes.
   */
  readBytes() {
    return Deno.readFile(this.path());
  }

  /**
   * @param to The absolute path to copy the file to.
   * @returns An unlocked source file of the copied file.
   */
  async copyTo(
    to: string,
    filepath = this.relativePath(),
  ): Promise<SourceFile> {
    const destination = join(to, filepath);

    if (destination === this.path()) {
      throw new Error(sprintf("cannot copy a file to itself: %s", this.path()));
    }

    await ensureDir(dirname(destination));
    await Deno.copyFile(this.path(), destination);

    const sourceFile = new SourceFile(destination, to);

    /**
     * This file has been copied, so we can safely unlock it.
     */
    sourceFile.unlock();

    return sourceFile;
  }

  async copyToHashed(to: string): Promise<SourceFile> {
    const contentHash = await this.contentHash();
    const extension = this.getExtension();

    const path = this.relativePath().replace(
      extension,
      `.${contentHash}${extension}`,
    );

    const copied = await this.copyTo(to, path);
    copied.aliasPath = join(to, this.relativePath());

    return copied;
  }

  isLocked() {
    return this.locked;
  }

  private unlock() {
    this.locked = false;

    return this;
  }

  async remove() {
    if (this.locked) {
      throw new Error(sprintf("cannot remove a locked file: %s", this.path()));
    }
    try {
      await Deno.remove(this.path());
      return true;
    } catch (error) {
      throw error;
    }
  }

  /**
   * @param length The length of the content hash to return.
   * @returns The hash of the files content.
   */
  async contentHash(length = 8) {
    const source = await this.readBytes();
    const hashBuffer = await crypto.subtle.digest("SHA-256", source);
    const hashArray = Array.from(new Uint8Array(hashBuffer));
    const hashHex = hashArray.map((b) => b.toString(16).padStart(2, "0")).join(
      "",
    );

    return hashHex.slice(0, length);
  }
}
