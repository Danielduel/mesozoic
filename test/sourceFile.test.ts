import { SourceFile } from "../lib/sources/sourceFile.ts";
import { assertEquals, assertRejects, assertSnapshot, join } from "./deps.ts";
import {
  ensureRelativePath,
  getFixtureDir,
  getFixturePath,
} from "./helpers.ts";

const outputDir = await Deno.makeTempDir();

function createSourceFile(path: string) {
  return new SourceFile(
    getFixturePath("app", path),
    getFixtureDir("app"),
  );
}

Deno.test("constructor", async () => {
  const path = join("src", "app.tsx");
  const sourceFile = createSourceFile(path);

  assertEquals(sourceFile.isLocked(), true);
  assertEquals(sourceFile.root(), getFixtureDir("app"));
  assertEquals(
    sourceFile.path(),
    getFixturePath("app", sourceFile.relativePath()),
  );
  assertEquals(sourceFile.relativePath(), ensureRelativePath(path));
  assertEquals(await sourceFile.contentHash(), "886415b0");
});

Deno.test("copyTo", async () => {
  const sourceFile = createSourceFile(join("src", "app.tsx"));

  const copied = await sourceFile.copyTo(outputDir);
  assertEquals(copied.root(), outputDir);
  assertEquals(copied.path(), join(outputDir, copied.relativePath()));
  assertEquals(copied.isLocked(), false);
  assertEquals(await copied.contentHash(), await sourceFile.contentHash());
  /**
   * We should be able to remove this file since it was copied
   */
  assertEquals(await copied.remove(), true);
});

Deno.test("copyToHashed", async () => {
  const sourceFile = createSourceFile(join("src", "app.tsx"));

  const copied = await sourceFile.copyToHashed(outputDir);
  assertEquals(copied.root(), outputDir);
  assertEquals(copied.path(), join(outputDir, copied.relativePath()));
});

Deno.test("remove a locked file", () => {
  const sourceFile = createSourceFile(join("src", "app.tsx"));
  assertRejects(() => sourceFile.remove());
});

Deno.test("read a file", async (t) => {
  const sourceFile = createSourceFile(join("src", "app.tsx"));
  assertSnapshot(t, await sourceFile.read());
});

Deno.test("read a non existent file", () => {
  const sourceFile = createSourceFile(join("src", "app2.tsx"));
  assertRejects(() => sourceFile.read());
});
