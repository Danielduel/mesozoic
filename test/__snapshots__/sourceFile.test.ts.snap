export const snapshot = {};

snapshot[`read a file 1`] = `
'import React from "https://esm.sh/react@18.2.0";

type AppProps = {
  name: string;
};

export default function App(props: AppProps) {
  return <div>Hello World</div>;
}
'
`;
